Transactions.md

```diff
@@ -1,182 +1,291 @@
-
-## <a name="ClosingUnilateral">Closing Transaction (Unilateral)</a>
-### <a name="ClosingKnownValues">Known Values</a>
-  * Local Final Address: `BitcoinAddress`
-  * nLockTime: `UInt32`
-  * Local Payout: `CurrencyUnit`
-  * ToLocalOutput: `ScriptPubKey`
-  * Fee Rate: `FeeUnit`
-
-Where
-  - `ToLocalOutput` is of the form [specified above](#CETOutputs)
-### <a name="ClosingGlobal">Global</a>
-  * nLockTime is `0`
-### <a name="ClosingInputs">Inputs</a>
-  * Input Spending(P2WSH(ToLocalOutput))
-      * nSequence is `0xffffffff`
-### <a name="ClosingOutputs">Outputs</a>
-  * One output corresponding to `Local Final Address` with value `Local Payout`
-
-## <a name="ClosingPenalty">Closing Transaction (Penalty)</a>
-### <a name="ClosingKnownValues">Known Values</a>
-  * Local Address: `BitcoinAddress`
-  * Remote's ToLocalOutput: `ScriptPubKey`
-  * Fee Rate: `FeeUnit`
-
-Where
-  - `Local Address` is any unused local address
-  - `Remote's ToLocalOutput` is of the form [specified above](#CETOutputs)
-### <a name="ClosingInputs">Inputs</a>
-  * Input Spending(P2WSH(Remote's ToLocalOutput))
-### <a name="ClosingOutputs">Outputs</a>
-  * One output corresponding to `LocalAddress` with value `P2WSH(Remote's ToLocalOutput).value - fee`
+This details the exact format of all on-chain transactions, which both sides need to agree on to ensure signatures are valid.
 
+# Table of Contents
+
+* [Funding Transaction](#funding-transaction)
+  * [Funding Transaction Input and Output Ordering](#funding-transaction-input-and-output-ordering)
+  * [Funding Inputs](#funding-inputs)
+  * [FundingOutput](#funding-output)
+  * [Change Outputs](#change-outputs)
+* [Contract Execution Transaction](#contract-execution-transaction)
+  * [CET Outputs](#cet-outputs)
+    * [Offerer Output](#offerer-output)
+    * [Accepter Output](#accepter-output)
+* [Refund Transaction](#refund-transaction)
+* [Fees](#fees)
+  * [Expected Weight of the Funding Transaction](#expected-weight-of-the-funding-transaction)
+  * [Expected Weight of the Contract Execution or Refund Transaction](#expected-weight-of-the-contract-execution-or-refund-transaction)
+* [Test Vectors](#test-vectors)
+* [References](#references)
+* [Authors](#authors)
+
+# Funding Transaction
+
+* version: 2
+* locktime: 0
+
+The funding inputs and change output script public keys are negotiated in the offer and accept messages.
+
+## Funding Transaction Input and Output Ordering
+
+The Offerer/Initiator's inputs come first and in the order that they were listed in the Offer message, the Accepter's inputs come after this in the order they are listed in the Accept message. The funding output comes first, followed by the Offerer's change output if applicable, and then the Accepter's change output if applicable.
+
+Note that this will likely change in the future.
+
+## Funding Inputs
+
+All funding inputs must be Segwit or nested P2SH(Segwit) in order to protect against malleability attacks.
```

Comment:
Even guaranteeing a party always has an anchor output to prevent witness-inflation pinning, I think the malicious party can just double-spend its announced outpoint with a high-fee, low-feerate transaction, stucking in network mempools and thus preventing confirmation of the funding.

If absolute fee of this transaction is higher than funding, it would be rejected by network mempools thus preventing any CPFP bump.

AFAICT, it's a pinning variant of the one LN is currently exposed to, low-feerate package based on concurrent states.

Reply:
What exactly is the downside of this? It seems that if this happens on the funding tx, the counter party can just double spend their outpoint the same to exit the transaction?

If your counterparty is being malicious during the funding process, you probably don't want to be in a contract with them?

Reply:
I have the same thought as @Christewart I don't really see the big issue here. You shouldn't consider the contract established until the fund transaction is confirmed.

Reply:
@Christewart, this happens post-funding process, when all signatures have been exchanged.

> It seems that if this happens on the funding tx, the counter party can just double spend their outpoint the same to exit the transaction?

In theory yes. In practice:
* Bob finds Alice full-node
* Bob mempools-split Alice's full-node from the rest of the network, feeding his high-fee, low-feerate double-spend to Alice, the funding_tx to others
* Bob's double spend doesn't propagate to the rest of network (just provoke a conflict on one of the input)
* Alice broadcast the funding tx, it's rejected _by her mempool_ as Bob's double spend doesn't signal RBF
* After few blocks, Alice try to double-spend her collateral input, her transaction is accepted locally, rejected by the rest of the network
* If DLC outcome isn't favorable to Bob, Bob wait until refund tx maturation and confirm funding_tx with a bump at that time

This assume full-node mapping and rapid feerate change to stuck the funding so it's assuming a lot on attacks resources. But AFAICT it formally works.

An good mitigation is to relative-timelock the refund tx to give opportunity to a honest party to commit CETs in case of malicious funding tx mempool-pinning.

Reply:
@ariard Thanks for the explanation, it's really interesting, but I still don't understand everything:
* How does having anchor output solves the problem here? If Alice's mempool rejected the fund transaction, how could she use her anchor output to bump the fees? Could you describe what she should do assuming that she notices the issue?
* How come is Alice's double spend on the fund transaction rejected by the rest of the network, while a potential double spend from Bob is not (I assume that's what bob would do if outcome is unfavorable?).

Reply:
From our last meeting, my understanding of @ariard explanation is that the main reason to enforce change output was to enable fee bumping a fund tx with CPFP without requiring to resign a tx which would have higher fee. Assuming that my understanding is correct, I would propose not in including this requirement in the first version of these specification since we have already so much to implement.